js_modules/dagit/packages/core/src/ui/Filters/FilterDropdown.tsx

```diff
@@ -23,7 +31,21 @@ interface FilterDropdownProps {
   setPortaledElements: React.Dispatch<React.SetStateAction<JSX.Element[]>>;
 }
 
+const focusedIndexAtom = atomFamily<number, string>({
+  key: 'FilterDropdown:focusedIndex',
+  default: -1,
+});
+const isFocusedSelector = selectorFamily<boolean, {key: string; index: number}>({
+  key: 'FilterDropdown:isFocused',
+  get: ({key, index}) => ({get}) => get(focusedIndexAtom(key)) === index,
+});
+
 export const FilterDropdown = ({filters, setIsOpen, setPortaledElements}: FilterDropdownProps) => {
+  const [menuKey, _] = React.useState(() => uuidv4());
+  const setFocusedItemIndex = useSetRecoilState(focusedIndexAtom(menuKey));
```

Comment:
It seems like this is a code smell. Making up UUIDs so we can shove component-local state into a global state controller? I guess I'm sorta with @hellendag that this seems like it should be basic local state.

Reply:
Optimizing renders in a list with a "focused" / "selected" state that changes on hover / interaction is actually a classic scenario where you don't want to use vanilla React state since it will cause more components to re-render than the ones that changed. Couple that with our extremely light usage of `useMemo` and `useCallback` and it's a recipe for re-rendering everything.

The other way to solve this (this = not re-rendering the whole list and only re-render the item losing focus and the item gaining focus) is to use context like I did here:  https://github.com/dagster-io/dagster/pull/10239

But that's uglier imo.

Reply:
The reason I want to prevent that render is to help callers of the component fall into the "Pit of success". Otherwise it's a bit like giving them a foot gun.

Reply:
That would have the extra re-renders issue mentioned before.  Solving it with just React requires Context that defines a custom hook for components to use to subscribe to only the state updates they care about (see this PR for an example: https://github.com/dagster-io/dagster/pull/10239) but I prefer this tbh.

Reply:
If you took this PR: https://github.com/dagster-io/dagster/pull/10239/files

And tried using basic React state instead, you'll get EXTREMELY sluggish and bad performance when hovering elements. And those components are not even complex. It's just the YAML config renderer.